Add delete routes

kmgreen2 opened the pull request:
I am not sure if others are hitting this, but I consistently got 405 whenever I tried issuing deletes using the official Golang GCS SDK.

Looking closer at the code, it seems like some of the routers are not in use?  In `buildMuxer()` (server.go:189), I see that `routers` is allocated, but never used.  Furthermore, I only see the routers defined on server.go:213 through server.go:225 when inspecting/debugging mux.go:136 (`mux:v1.8.0`).

What is the intention of `server.go:198`?  Should that be removed, or am I missing a side-effect.  If there is a side-effect associated with routers, I would suggest you directly define those routes.

In any case, deletes are returning a 405 because the server cannot find a route.  Since I am not sure of the intention of `server.go:198`, I have added new delete routes directly to `s.mux`, which fixes the issue.

fsouza commented:
Hi @kmgreen2, object deletion should work (bucket deletion is not implemented, see #214). The routers in the slice are defined as subrouters of the main router and that's where the side effect happens. Can probably add a comment or something like that, but that's unfortunately how gorilla/mux works.

Looking at your diff, it seems that your code is trying to delete objects without the `/storage/v1` prefix? Which SDK are you using? Can you share a reproducer?

fsouza commented:
Oh I can't read, you mentioned you were using the Go SDK. I can give it a shot later.

fsouza commented:
Can't dig into this right now, but opened #441 that should provide a reproducer.

If that PR passes, could you share some code to reproduce the behavior you're seeing?

kmgreen2 commented:
@fsouza Thanks for the quick reply!

It is possible I am not starting the server properly (I just started using your library yesterday :) )

Here is how I am starting the GCS server (I had to hijack DNS resolution, so I can include this in tests without updating `/etc/hosts`):

```
func SetupGCSLocal() (*fakestorage.Server, error){
	return fakestorage.NewServerWithOptions(fakestorage.Options{
		InitialObjects: []fakestorage.Object{
			{
				BucketName: "localtest",
				Name:       "some/object/foo",
				Content:    []byte("bar"),
			},
		},
		Scheme: "http",
		Host: "localhost",
		Port: 8081,
		Writer: os.Stdout,
		PublicHost: "localtest.localhost",
	})
}
```

I then create a new client using the following.  It took some trial and error to get the SDK to work properly locally, which appears to be more of a Google SDK problem than anything to do with fake-gcs-server:

```
os.Setenv("STORAGE_EMULATOR_HOST", "localhost:8081")
storage.NewClient(ctx, 
    option.WithoutAuthentication(),
    option.WithHTTPClient(&http.Client{Transport: GetFakeTransport()}),
    option.WithEndpoint("http://localtest.localhost:8081"))
```

Thanks, again!

kmgreen2 commented:
@fsouza I see that the tests do not rely on running the fake GCS daemon, could that be related?

kmgreen2 commented:
I took a closer look at your examples and noticed that they only include list and download (and delete with your PR).  The code snippets above were definitely made much more complicated to support upload locally.  I tried to reproduce with upload and delete using `examples/`, but am unable to build the server locally or with Docker:

```
docker build -t fsouza/fake-gcs-server .
Sending build context to Docker daemon  275.5kB
Step 1/12 : FROM golang:1.16.0 AS builder
 ---> f15d23d96763
Step 2/12 : WORKDIR /code
 ---> Using cache
 ---> 31cf0090e68b
Step 3/12 : ADD go.mod go.sum ./
 ---> Using cache
 ---> 9e7876ffa5c1
Step 4/12 : RUN go mod download
 ---> Using cache
 ---> a65474170c25
Step 5/12 : ADD . ./
 ---> Using cache
 ---> 6d4d46dfb0f9
Step 6/12 : ENV CGO_ENABLED=0
 ---> Using cache
 ---> db56529fd654
Step 7/12 : RUN go build -o fake-gcs-server
 ---> Running in a1883d71be1b
package github.com/kmgreen2/fake-gcs-server
	main.go:18:2: use of internal package github.com/fsouza/fake-gcs-server/internal/config not allowed
The command '/bin/sh -c go build -o fake-gcs-server' returned a non-zero code: 1
```

In any case, my PR works for upload, download, delete and head, so I am unblocked by overloading go.mod with my branch.  Since this is just for tests, it should be fine for now.

kmgreen2 commented:
Maybe part of my problem is that I am setting up my client to upload, download and delete, which requires more than your examples provide.  I added an upload step to your Go example and it does not work properly with the official Google Go SDK.

I added this helper:
```
func uploadFile(client *storage.Client, bucketName, key string, reader io.Reader) error {
    wc := client.Bucket(bucketName).Object(key).NewWriter(context.Background())
    if _, err := io.Copy(wc, reader); err != nil {
        return fmt.Errorf("io.Copy: %v", err)
    }
    if err := wc.Close(); err != nil {
        return fmt.Errorf("Writer.Close: %v", err)
    }
    return nil
}```

and called it after `downloadFile` in `example/go/main.go`:

```
   payloadBytes := bytes.NewBuffer([]byte("foobar"))
    err = uploadFile(client, bucketName, "uploadTestKey", payloadBytes)
    if err != nil {
        log.Fatal(err)
    }
    fmt.Printf("uploaded %s to %s/uploadTestKey\n", string(payloadBytes.Bytes()), bucketName)
```

I get this result:

```
$ go run main.go
buckets: [some_file.txt]
contents of sample-bucket/some_file.txt: Some amazing content to be loaded
2021/03/02 16:11:59 Writer.Close: Post "http://http/upload/storage/v1/b/sample-bucket/o?alt=json&name=uploadTestKey&prettyPrint=false&projection=full&uploadType=multipart": dial tcp: lookup http: no such host
exit status 1
```

Doing what I did above seems to work for upload, delete, download and head.  It seems that the official Google SDK doesn't like local instances of GCS, so requires a bit of extra configuration.

In any case, I am unblocked by using my PR and goofy client-side configuration, but would be happy to help if you think this is a general problem.

kmgreen2 commented:
@fsouza FYI, I unblocked myself with this PR in my fork.  It took a bit of wrangling to get the Google SDK working for all endpoints locally.  If you think there is an actual problem here, I am happy to help, but understand if you don't.  I can just pin to my fork, since this represents a small number of tests in my project (https://github.com/kmgreen2/agglo/blob/main/pkg/storage/gcsobj_test.go) and I have them working with my fork.
